src/System.Memory/src/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed partial class ArrayMemoryPool<T> : MemoryPool<T>
+    {
+        private sealed class ArrayMemoryPoolBuffer : OwnedMemory<T>
+        {
+            private T[] _array;
+            private bool _disposed;
+            private int _referenceCount;
+
+            public ArrayMemoryPoolBuffer(int size)
+            {
+                _array = ArrayPool<T>.Shared.Rent(size);
+            }
+
+            public sealed override int Length => _array.Length;
+
+            public sealed override bool IsDisposed => _disposed;
+
+            protected sealed override bool IsRetained => _referenceCount > 0;
+
+            public sealed override Span<T> Span
+            {
+                get
+                {
+                    if (IsDisposed)
+                        ThrowHelper.ThrowObjectDisposedException(nameof(ArrayMemoryPoolBuffer));
+
+                    return _array;
+                }
+            }
+
+            protected sealed override void Dispose(bool disposing)
+            {
+                T[] array = Interlocked.Exchange(ref _array, null);
+                if (array != null)
+                {
+                    _disposed = true;
+                    ArrayPool<T>.Shared.Return(array);
+                }
+            }
+
+            protected
+#if netstandard // TryGetArray is exposed as "protected internal". Normally, the rules of C# dictate we override it as "protected" because the base class is
+                // in a different assembly. Except in the netstandard config where the base class is in the same assembly.
+            internal
+#endif
+            sealed override bool TryGetArray(out ArraySegment<T> arraySegment)
+            {
+                if (IsDisposed)
+                    ThrowHelper.ThrowObjectDisposedException(nameof(ArrayMemoryPoolBuffer));
+
+                arraySegment = new ArraySegment<T>(_array);
+                return true;
+            }
+
+            public sealed override MemoryHandle Pin(int byteOffset = 0)
+            {
+                unsafe
+                {
+                    Retain(); // this checks IsDisposed
+                    GCHandle handle = GCHandle.Alloc(_array, GCHandleType.Pinned);
```

Comment:
Add argument checks for byteOffset?

Similar to: https://github.com/dotnet/corefx/blob/master/src/System.Memory/tests/Memory/CustomMemoryForTest.cs#L47
```C#
if (byteOffset < 0 || (byteOffset/Unsafe.SizeOf<T>()) > _array.Length) throw
```

Reply:
This check isn't right - for an array of 10 ints, "40" should be a valid input, but not 41, 42 and 43.

This seems like low-level api to be doing this kind of check - I'll put in but it gets tricky since MemoryPool owners don't have strict control over the sizes they get back. The byteOffset might not actually fit in an Int32.